lightningd/bitcoind.c

```diff
@@ -141,189 +55,239 @@ static void bcli_failure(struct bitcoind *bitcoind,
+	       (arg[0] == '"' && arg[arglen - 1] == '"');
 }
 
+static void add_input(char **cmd, const char *input, bool last)
+{
+	/* Numbers, bools, objects and arrays are left unquoted,
+	 * and quoted things left alone. */
+	if (is_literal(input))
+		tal_append_fmt(cmd, "%s", input);
+	else
+		tal_append_fmt(cmd, "\"%s\"", input);
+	if (!last)
+		tal_append_fmt(cmd, ", ");
+}
+
+static bool process_donothing(struct bitcoin_rpc *bcrpc UNUSED) { return true; }
+
 /* If stopper gets freed first, set process() to a noop. */
-static void stop_process_bcli(struct bitcoin_cli **stopper)
+static void stop_process_brpc(struct bitcoin_rpc **stopper)
 {
 	(*stopper)->process = process_donothing;
 	(*stopper)->stopper = NULL;
 }
 
-/* It command finishes first, free stopper. */
-static void remove_stopper(struct bitcoin_cli *bcli)
+/* It rpc command finishes first, free stopper. */
+static void remove_stopper(struct bitcoin_rpc *brpc)
 {
-	/* Calls stop_process_bcli, but we don't care. */
-	tal_free(bcli->stopper);
+	/* Calls stop_process_brpc, but we don't care. */
+	tal_free(brpc->stopper);
 }
 
-/* If ctx is non-NULL, and is freed before we return, we don't call process().
- * process returns false() if it's a spurious error, and we should retry. */
-static void
-start_bitcoin_cli(struct bitcoind *bitcoind,
-		  const tal_t *ctx,
-		  bool (*process)(struct bitcoin_cli *),
-		  bool nonzero_exit_ok,
-		  enum bitcoind_prio prio,
-		  void *cb, void *cb_arg,
-		  char *cmd, ...)
+static struct bitcoin_rpc *
+start_bitcoin_rpc(struct bitcoind *bitcoind, const tal_t *ctx,
+		  bool (*process)(struct bitcoin_rpc *), bool rpc_error_ok,
+		  enum bitcoind_prio prio, void *cb, void *cb_arg, char *cmd,
+		  ...)
 {
 	va_list ap;
-	struct bitcoin_cli *bcli = tal(bitcoind, struct bitcoin_cli);
-
-	bcli->bitcoind = bitcoind;
-	bcli->process = process;
-	bcli->prio = prio;
-	bcli->cb = cb;
-	bcli->cb_arg = cb_arg;
+	struct bitcoin_rpc *brpc = tal(ctx, struct bitcoin_rpc);
+	const char *arg, *next_arg;
+
+	brpc->bitcoind = bitcoind;
+	brpc->process = process;
+	brpc->cb = cb;
+	brpc->cb_arg = cb_arg;
+	brpc->prio = prio;
+	brpc->exitstatus = RPC_FAIL;
+	brpc->rpc_error_ok = rpc_error_ok;
+	brpc->resulttok = NULL;
+	brpc->errortok = NULL;
+	brpc->errorcode = 0;
 	if (ctx) {
 		/* Create child whose destructor will stop us calling */
-		bcli->stopper = tal(ctx, struct bitcoin_cli *);
-		*bcli->stopper = bcli;
-		tal_add_destructor(bcli->stopper, stop_process_bcli);
-		tal_add_destructor(bcli, remove_stopper);
+		brpc->stopper = tal(ctx, struct bitcoin_rpc *);
+		*brpc->stopper = brpc;
+		tal_add_destructor(brpc->stopper, stop_process_brpc);
+		tal_add_destructor(brpc, remove_stopper);
 	} else
-		bcli->stopper = NULL;
+		brpc->stopper = NULL;
+
+	brpc->cmd = tal_fmt(brpc, "%s ", cmd);
+	brpc->request = tal_fmt(brpc,
+				"{\"jsonrpc\": \"1.0\", \"id\":\"lightningd\", "
+				"\"method\": \"%s\", \"params\":",
+				cmd);
+
+	tal_append_fmt(&brpc->request, "[ ");
```

Comment:
We have a library built on top of the `json_request` struct that should facilitate the creation of JSON requests. That makes this kind of operations easier.